frontend/src/app/matching/matching.component.ts

```diff
@@ -1,72 +1,113 @@
-import { Component } from '@angular/core'
+import { Component, OnDestroy, OnInit } from '@angular/core'
 import { NgForm } from '@angular/forms'
 import { Router } from '@angular/router';
 import { MatchingService } from '../_services/matching.service'
 import { StorageService } from '../_services/storage.service'
-import { timeoutWith, throwError } from 'rxjs'
+import { timeoutWith, throwError, count, interval, range, timer, takeWhile, scan } from 'rxjs'
 
 @Component({
   selector: 'app-matching',
   templateUrl: './matching.component.html',
   styleUrls: ['./matching.component.css']
 })
-export class MatchingComponent {
+export class MatchingComponent implements OnInit, OnDestroy {
 
   match: any = undefined
 
   requested: boolean = false
 
   queueLength: number = -1
 
+  currUser = this.storageService.getUser().username
+
+  timer = {s: 0};
+
+  countDown: any = undefined
+
+  enqueueSubscription: any = undefined
+
   constructor(
     private router: Router,
     private matchingService: MatchingService,
     private storageService: StorageService){}
 
   ngOnInit() {
-    this.getQueueLength()
+    // this.getQueueLength()
+  }
+
+  ngOnDestroy(): void {
+      this.cancelMatch()
   }
 
   getMatch(userDetails: any) {
-    this.matchingService.enqueue(userDetails).pipe(timeoutWith(30000, throwError(() => {
-      this.matchingService.dequeue(userDetails["userid"]).subscribe((res) => {
+    if (this.countDown) {
+      this.countDown.unsubscribe();
+    }
+    this.countDown = timer(0, 1000).pipe(
+      scan(i => --i, 31),
+      takeWhile(i => i >= 0 && this.requested)
+    ).subscribe((i: number) => {
+      this.timer.s = 30 - i;
+      if (this.requested) {
+        document.getElementById("bar")!.style.width = (this.timer.s/30)*100+"%";
+        document.getElementById("bar")!.innerHTML = this.timer.s + "s";
+      }
+    });
+    this.enqueueSubscription = this.matchingService.enqueue(userDetails)
+    .pipe(timeoutWith(30000, throwError(() => {
+      this.matchingService.dequeue(this.currUser).subscribe((res) => {
         this.match = "Your request timed out!"
-        this.requested = false
-        this.getQueueLength()
+        setTimeout(() => this.requested = false, 5000);
+        // this.getQueueLength()
       })
     }))).subscribe((res) => {
       if (res.message.includes("Matched users:")) {
         this.match = res.message
-        this.getQueueLength()
+        // this.getQueueLength()
+        this.matchingService.updateMatchId(res.match);
         this.router.navigate(["/collab", res.roomId, res.difficulty, res.language]);
+      } else {
+        // this.getQueueLength();
+      }
+    }, (err) => {
+      if (err) {
+        this.match = "We encountered a problem. Please try again later."
+        // this.getQueueLength()
+        this.cancelMatch()
       }
     })
-    this.getQueueLength()
+    // this.getQueueLength()
```

Comment:
Remove previous queue length artifact comments

Reply:
Removed in most recent push.